src/wallet/walletutil.cpp

```diff
@@ -25,3 +25,46 @@ fs::path GetWalletDir()
 
     return path;
 }
+
+static bool IsBerkeleyBtree(const fs::path& path)
+{
+    // Berkeley DB Btree magic bytes, from:
+    //  https://github.com/file/file/blob/5824af38469ec1ca9ac3ffd251e7afe9dc11e227/magic/Magdir/database#L74-L75
+    const uint32_t bdb_magic = 0x00053162;
+
+    std::ifstream file(path.string(), std::ios::binary);
+    if (!file.is_open()) return false;
+
+    file.seekg(12, std::ios::beg); // Magic bytes start at offset 12
+    uint32_t data = 0;
+    file.read((char*) &data, sizeof(data)); // Read 4 bytes of file to compare against magic
+
+    return data == bdb_magic;
+}
+
+std::vector<fs::path> ListWalletDir()
+{
+    const fs::path wallet_dir = GetWalletDir();
+    std::vector<fs::path> paths;
+
+    for (auto it = fs::recursive_directory_iterator(wallet_dir); it != end(it); ++it) {
+        if (it->status().type() == fs::directory_file && IsBerkeleyBtree(it->path() / "wallet.dat")) {
+            // Found a directory which contains wallet.dat btree file, add it as a wallet.
+            paths.emplace_back(fs::relative(it->path(), wallet_dir));
+        } else if (it.level() == 0 && it->symlink_status().type() == fs::regular_file && IsBerkeleyBtree(it->path())) {
```

Comment:
No, it only recursively scans directories to try to open `wallet.dat` inside and then check magic bytes.

It checks magic bytes on all files in the walletdir (datadir in the legacy case).

Reply:
It still might be good to prevent this from opening too many files if there happen to be a lot of stray files in the root directory, or a lot of directories in the tree. Maybe add `&& (++count)` before `&& IsBerkeleyBtree` checks, and `if (count > 1000) throw std::runtime_error("Can't list wallets, too many files in walletdir")` and the end of the loop as a sanity check.

Reply:
@ryanofsky an alternative solution is to scan only if walletdir != datadir, WDYT?

Reply:
> It doesn't provide protection if there's a looping directory structure

From the [docs](https://www.boost.org/doc/libs/1_67_0/libs/filesystem/doc/reference.html#Class-recursive_directory_iterator):
> By default, recursive_directory_iterator does not follow directory symlinks

IIUC considering the above looping would't happen.

Reply:
> By default, recursive_directory_iterator does not follow directory symlinks

I know, I was thinking about more about FUSE mounts and such. Anyway, if you think it complicates code too much to deal with unexpected filesystem stuff and provide a nice error message, I don't personally think you need to add any error handling here. I was just trying to suggest a way to deal with laanwj's concern.

But since you asked WDYT, I think special casing datadir=walletdir is a bad user experience and confusing and ugly and hard to explain.